.buildkite/scripts/build-pipeline/generate-steps.py

```diff
@@ -0,0 +1,101 @@
+import os
+import requests
+import sys
+import typing
+from requests.adapters import HTTPAdapter, Retry
+
+from ruamel.yaml import YAML
+
+RELEASES_URL = "https://raw.githubusercontent.com/elastic/logstash/main/ci/logstash_releases.json"
+TEST_COMMAND: typing.final = ".buildkite/scripts/run_tests.sh"
+
+
+def generate_unit_and_integration_test_steps(stack_version, snapshot) -> list[typing.Any]:
+    test_steps = []
+
+    # step-1, unit tests
+    label_unit_test: typing.final = f"Unit test for {stack_version}, snapshot: {snapshot}"
+    test_steps.append({
+        "label": label_unit_test,
+        "command": TEST_COMMAND,
+        "env": {
+            "SNAPSHOT": snapshot,
+            "ELASTIC_STACK_VERSION": stack_version,
+            "INTEGRATION": "false"
+        }
+    })
+
+    # step-2, integration tests
+    label_integration_test: typing.final = f"Integration test for {stack_version}, snapshot: {snapshot}"
+    test_steps.append({
+        "label": label_integration_test,
+        "command": TEST_COMMAND,
+        "env": {
+            "SNAPSHOT": snapshot,
+            "ELASTIC_STACK_VERSION": stack_version,
+            "INTEGRATION": "true",
+            "SECURE_INTEGRATION": "true",
+            "LOG_LEVEL": "info"
+        }
+    })
+    return test_steps
+
+
+def call_url_with_retry(url: str, max_retries: int = 5, delay: int = 1) -> requests.Response:
+    schema = "https://" if "https://" in url else "http://"
+    session = requests.Session()
+    # retry on most common failures such as connection timeout(408), etc...
+    retries = Retry(total=max_retries, backoff_factor=delay, status_forcelist=[408, 502, 503, 504])
+    session.mount(schema, HTTPAdapter(max_retries=retries))
+    return session.get(url)
+
+
+if __name__ == "__main__":
+    structure = {
+        "agents": {
+            "provider": "gcp",
+            "machineType": "n1-standard-4",
+            "image": "family/core-ubuntu-2204"
+        },
+        "steps": []}
+
+    steps = []
+    response = call_url_with_retry(RELEASES_URL)
+    versions_json = response.json()
+
+    # there are situations to manually run CIs with PR change,
+    # set MANUAL_TARGET_BRANCH with upstream target branch and run
+    manually_set_target_branch: typing.final = os.getenv("MANUAL_TARGET_BRANCH")
+    target_branch: typing.final = manually_set_target_branch if manually_set_target_branch else os.getenv("TARGET_BRANCH")
```

Comment:
Can we print or log out what target_branch is resolved to? That may help with debugging especially if we are relying on env vars being set.

Reply:
Its a bit strange the way the `TARGET_BRANCH` appears to be set in the manual workflow. In a pull request there is a `source` branch which contains the changes to be merged into a `target` branch. In the manual workflow scenaro the `TARGET_BRANCH` is set (presumably because there is no way to resolve a "source" branch). Probably the manual workflow is not all that common and documenting it will be fine. But any chance we can make sure to keep naming consistency around `target` and `source` to distinguish between the branch that caries the changes and the branch we are intending to merge in to would be helpful.